src/traffic_server/traffic_server.cc

```diff
@@ -452,6 +453,33 @@ class MemoryLimit : public Continuation
   struct rusage _usage;
 };
 
+/** Gate the emission of the "Traffic Server is fuly initialized" log message.
+ *
+ * This message is intended to be helpful to users who want to know that
+ * Traffic Server is not just running but has become fully initialized and is
+ * ready to optimize traffic. This is in contrast to the "traffic server is
+ * running" message which can be printed before either of these conditions.
+ *
+ * This function is called on each initialization state transition. Currently,
+ * the two state transitions of interest are:
+ *
+ * 1. The cache is initialized.
+ * 2. The ports are open and accept has been called upon them.
```

Comment:
Do you mean listen rather than accept?

Reply:
I actually do mean accept. 

Here's some more context: Alan showed me that the ports may be opened and listen may have been called on the ports before cache is initialized and before accept is called on them, and that the difference in time between the opening of the ports and the accept can be large (over 10 ms, say).

Here's the sequence of events that Traffic Server follows with regard to configuring its ports:

1. Traffic server initializes data structures for the ports it listens on. It may bind upon those ports depending upon Traffic Server configuration. (Our docs say as much, see https://docs.trafficserver.apache.org/en/9.0.x/developer-guide/api/functions/TSLifecycleHookAdd.en.html?highlight=life%20cycle#_CPPv4N17TSLifecycleHookID35TS_LIFECYCLE_PORTS_INITIALIZED_HOOKE)
1. Traffic Server calls TS_LIFECYCLE_PORTS_INITIALIZED_HOOK on all plugins registered for it so they can do whatever they want to do to prepare for connections.
1. If wait_for_cache is called, we delay calling accept on the ports until cache is initialized, otherwise we directly...
1. Call start_HttpProxyServer(), which calls accept on the ports.

Before this patch, the AuTest extension considered Traffic Server ready to accept test traffic if it succeeded in opening a connection to the listening port. But this could happen in step one above, before cache is initialized and before accept is called on the ports, **even if wait_for_cache is configured**. This change adds a message to wait upon a message that is designed to guarantee that everything is done and optimization is ready.

Reply:
I'm adding a paragraph to flesh this out further. Let me know if it is helpful.

Reply:
Sorry, my bad, somehow I thought for TCP, listen blocked until it got the first SYN.